### PR TITLE
EIP-4527: QR Code data transmission protocol for the Ethereum wallets

### PR DESCRIPTION
**Pull request by aaronisme**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: EIPS/eip_draft_qr_code_data_transmission_protocol.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
+---
+eip: eip_draft_qr_code_data_transmission_protocol
```
Comment (lightclient) [inline EIPS/eip_draft_qr_code_data_transmission_protocol.md:2]:
```suggestion
eip: 4527
```

File: EIPS/eip_draft_qr_code_data_transmission_protocol.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+Because of these advantages, the QR Code data transmission is a better choice. But currently, there is no standard for how the offline signer works with the watch-only wallet and how the data should be encoded.
+This EIP presents a standard process and data transmission protocol for offline signers to work with the watch-only wallet.
+
+## Specification
+**Offline signer**: the offline signer is a device or application which holds the user’s private keys and does not have network access.
+
+**Watch-only wallet**: the watch-only wallet is the wallet that has network access and will interact with the Ethereum blockchain.
+
+### Process:
```
Comment (lightclient) [inline EIPS/eip_draft_qr_code_data_transmission_protocol.md:32]:
```suggestion
### Process
```

File: EIPS/eip_draft_qr_code_data_transmission_protocol.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+## Reference Implementation
+The reference implementation is in Javascript and available at [https://github.com/KeystoneHQ/keystone-airgaped-base/tree/master/packages/ur-registry-eth]
+
+Metamask has adopted it for the integration with QR-based Signer. https://github.com/MetaMask/metamask-extension/pull/12065
+
+Here is the video to show how it works: https://www.youtube.com/watch?v=1eM53TYG1YA
+
+## Security Considerations
+The offline signer should decode all the data from `eth-sign-request` and show them to the user to confirm before signing. It is recommended to provide an address field in the `eth-sign-request`, If it is provided, the offline signer should verify the address is the same as the signing key associated address.
+
```
Comment (lightclient) [inline EIPS/eip_draft_qr_code_data_transmission_protocol.md:213]:
```suggestion
```

File: EIPS/eip_draft_qr_code_data_transmission_protocol.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
+---
+eip: eip_draft_qr_code_data_transmission_protocol
+title: QR Code data transmission protocol for the Ethereum wallets
+description: QR Code data transmission protocol between wallets and offline signers.
+author: Aaron Chen (@aaronisme), Sora Lee (@soralit), ligi(@ligi), Dan Miller(@danjm), AndreasGassmann (@andreasgassmann), xardass(@xardass), Lixin Liu (@BitcoinLixin)
```
Comment (lightclient) [inline EIPS/eip_draft_qr_code_data_transmission_protocol.md:5]:
```suggestion
author: Aaron Chen (@aaronisme), Sora Lee (@soralit), ligi (@ligi), Dan Miller (@danjm), AndreasGassmann (@andreasgassmann), xardass (@xardass), Lixin Liu (@BitcoinLixin)
```

**Review by lightclient — CHANGES_REQUESTED** [review]

**Comment by alita-moore** [comment]
> A critical exception has occurred (cc @alita-moore):
> 	Message: incomplete explicit mapping pair; a key node is missed; or followed by a non-tabulated empty line at line 4, column 15:
>     author: author: aaron chen (@aaronisme), sora  ... 
>                   ^
> 

There is a ticket in place to address this unhandled exception

**Comment by aaronisme** [comment]
> > A critical exception has occurred (cc @alita-moore):
> > Message: incomplete explicit mapping pair; a key node is missed; or followed by a non-tabulated empty line at line 4, column 15:
> > author: author: aaron chen (@aaronisme), sora  ...
> > ^
> 
> There is a ticket in place to address this unhandled exception

The issue have been fixed

**Comment by aaronisme** [comment]
@lightclient I have fixed the comments you mentioned and renamed the file as `eip-4527.md`. Would you please do another review ?

**Comment by alita-moore** [comment]
@aaronisme yep, so this is a bot and there are certain situations it's not equipped to handle. I was making it known that I will be equipping the bot someone soon.

**Comment by aaronisme** [comment]
> @aaronisme yep, so this is a bot and there are certain situations it's not equipped to handle. I was making it known that I will be equipping the bot soon.

Sure, Thanks

**Comment by BitcoinLixin** [comment]
To whom it may concern,

This EIP has been implemented to Keystone hardware wallet & MetaMask extension wallet integration - https://twitter.com/MetaMask/status/1469351219290382341
Demo Video - https://youtu.be/1eM53TYG1YA

Currently,
* [Airgap.it](https://airgap.it/) is following this EIP to integrate with MetaMask extension wallet. This app can turn people's old mobile phone into an offline signer with QR code.
* Keystone team is working on MetaMask Mobile wallet integration following this EIP.
* [Rabby Wallet](https://rabby.io/) is integrating Keystone following this EIP.

**Review by lightclient — APPROVED** [review]
Overall looks good, just a few comments.

File: EIPS/eip-4527.md [hunk 5]
```diff
@@ -0,0 +1,212 @@
+---
+eip: 4527
+title: QR Code data transmission protocol for the Ethereum wallets
```
[the remark this answers is not in the document]
Comment (aaronisme) [reply EIPS/eip-4527.md]: Updated

File: EIPS/eip-4527.md [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+1. The offline signer provides the public key information to the watch-only wallet to generate addresses, sync balances and etc via QR Codes.
+2. The watch-only wallet generates the unsigned data and sends it to an offline signer for signing via QR Code, data that can include transactions, typed data, and etc.
+3. The offline signer signs the data and provides a signature back to the watch-only wallet via QR Code.
+4. The watch-only wallet receives the signature, constructs the signed data (transaction) and performs the following activities like broadcasting the transaction etc.
+
+### Data Transmission Protocol
+
+Since a single QR Code can only contain a limited amount of data, animated QR Codes should be utilized for data transmission. The [BlockchainCommons](https://github.com/BlockchainCommons/) have published a series of data transmission protocol called Uniform Resources (UR). It provides a basic method to encode data into animated QR Codes. This EIP will use UR and extend its current definition. For more info about UR, please check it out [here](https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-005-ur.md).
```
[the remark this answers is not in the document]
Comment (aaronisme) [reply EIPS/eip-4527.md]: Sure, updated the link and add the commit

**Pull request merged**